src/NuGet.Core/NuGet.Protocol.Core.v3/Utility/MetadataReferenceCache.cs

```diff
@@ -0,0 +1,146 @@
+        /// If not, caches the parsed value and returns it.
+        /// </summary>
+        public NuGetVersion GetVersion(string s)
+        {
+            NuGetVersion version;
+            if (!NuGetVersion.TryParse(s, out version))
+            {
+                return null;
+            }
+
+            return GetVersion(version);
+        }
+
+        /// <summary>
+        /// Checks if <paramref name="version"/> already exists in the cache.
+        /// If so, returns the cached instance.
+        /// If not, caches <paramref name="version"/> and returns it.
+        /// </summary>
+        public NuGetVersion GetVersion(NuGetVersion version)
+        {
+            NuGetVersion cachedVersion;
+            if (!_versionCache.TryGetValue(version, out cachedVersion))
+            {
+                var systemVersion = GetSystemVersion(version);
+
+                // Save memory by caching the System.Version and the release label string separately.
+                var releaseLabels = version.ReleaseLabels.Select(GetString);
+
+                // Rebuild the version without the original string value.
+                version = new NuGetVersion(systemVersion, releaseLabels, GetString(version.Metadata), originalVersion: null);
+
+                _versionCache.Add(version, version);
+                cachedVersion = version;
+            }
+
+            return cachedVersion;
+        }
+
+        private Version GetSystemVersion(NuGetVersion nugetVersion)
+        {
+            var version = new Version(nugetVersion.Major, nugetVersion.Minor, nugetVersion.Patch, nugetVersion.Revision);
+
+            Version cachedVersion;
+            if (!_systemVersionCache.TryGetValue(version, out cachedVersion))
+            {
+                _systemVersionCache.Add(version, version);
+                cachedVersion = version;
+            }
+
+            return cachedVersion;
+        }
+
+        /// <summary>
+        /// Mapping of input parameter type to caching method.
+        /// </summary>
+        private static readonly IDictionary<Type, string> CachableTypesMap = new Dictionary<Type, string>
+        {
+            {typeof(string), nameof(GetString)},
+            {typeof(NuGetVersion), nameof(GetVersion)}
+        };
+
+        /// <summary>
+        /// <see cref="IEnumerable{Type}"/> containing all types that can be cached.
+        /// </summary>
+        public static IEnumerable<Type> CachableTypes => CachableTypesMap.Keys;
+        
+        /// <summary>
+        /// Iterates through the properties of <paramref name="input"/> that are either <see cref="string"/>s, <see cref="DateTimeOffset"/>s, or <see cref="NuGetVersion"/>s and checks them against the cache.
+        /// </summary>
+        public T GetObject<T>(T input)
+        {
+            // Get all properties that contain both a Get method and a Set method and can be cached.
+            var properties =
+                typeof(T).GetTypeInfo()
```

Comment:
I finished testing and wasn't able to see any major perf improvements or degradations. I think this is partly caused by the client making mostly single package requests. Since my changes only optimize at a per request level, it likely doesn't make much of an impact. I think it would be fairly easy to make this cache more widely used and increase its usefulness.